packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -2313,18 +2367,81 @@ Hardhat Network's forking functionality only works with blocks from at least spu
-  try {
-    const common = new Common({ chain: networkId });
+  private _selectHardfork(
+    blockNumber: BN,
+    defaultName = this._vm._common.hardfork() as HardforkName
+  ): string {
+    if (
+      this._forkBlockNumber === undefined ||
+      blockNumber.gte(new BN(this._forkBlockNumber))
+    ) {
+      return defaultName;
+    }
 
-    common.setHardfork(common.activeHardfork(blockNumber));
+    if (this._hardforkActivations.size === 0) {
+      throw new InternalError(
+        `No known hardfork for execution on historical block ${blockNumber.toString()} (relative to fork block number ${
+          this._forkBlockNumber
+        }). The node was not configured with a hardfork activation history.  See http://hardhat.org/hardhat-network/guides/mainnet-forking.html#using-a-custom-hardfork-history`
+      );
+    }
 
-    return common;
-  } catch (e) {
-    throw new InternalError(
-      `Network id ${networkId} does not correspond to a network that Hardhat can trace`
+    /** search this._hardforkActivations for the highest block number that
+     * isn't higher than blockNumber, and then return that found block number's
+     * associated hardfork name. */
+    const hardforkHistory: Array<[name: string, block: number]> = Array.from(
+      this._hardforkActivations.entries()
     );
+    const [hardfork] = hardforkHistory.reduce(
+      ([highestHardfork, highestBlock], [thisHardfork, thisBlock]) =>
+        thisBlock > highestBlock && new BN(thisBlock).lte(blockNumber)
+          ? [thisHardfork, thisBlock]
+          : [highestHardfork, highestBlock]
+    );
+    if (hardfork === undefined) {
+      throw new InternalError(
+        `Could not find a hardfork to run for block ${blockNumber}, after having looked for one in the HardhatNode's hardfork activation history, which was: ${JSON.stringify(
+          hardforkHistory
+        )}. For more information, see https://hardhat.org/hardhat-network/reference/#config`
+      );
+    }
+
+    if (!HARDHAT_NETWORK_SUPPORTED_HARDFORKS.includes(hardfork)) {
+      throw new InternalError(
+        `Tried to run a call or transaction in the context of a block whose hardfork is "${hardfork}", but Hardhat Network only supports the following hardforks: ${HARDHAT_NETWORK_SUPPORTED_HARDFORKS.join(
+          ", "
+        )}`
+      );
+    }
+
+    return hardfork;
+  }
+
+  private _getCommonForTracing(networkId: number, blockNumber: number): Common {
+    try {
+      const common = new Common({
+        chain: {
+          // eslint-disable-next-line @typescript-eslint/dot-notation
+          ...Common["_getChainParams"]("mainnet"),
+          chainId: networkId,
+          networkId,
+        },
+        hardfork: this._selectHardfork(new BN(blockNumber)),
+      });
+
+      common.setHardfork(
+        this._selectHardfork(
+          new BN(blockNumber),
+          common.activeHardfork(blockNumber) as HardforkName
+        )
+      );
```

Comment:
This is duplicated, as the hardfork was already set in the constructor. Also, i'll feel more comfortable if we don't mutate custom commons, as that has lead to weird results before.

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/1928/commits/bdcd35e3428a7ab79256eb9661be0a5bfcc0c83c